cmd/oceanbench/set.go

```diff
@@ -368,6 +370,149 @@ func editDevicesHandler(w http.ResponseWriter, r *http.Request) {
+	}
+
+	mac := r.FormValue("ma")
+	vb, err := strconv.ParseFloat(r.FormValue("vb"), 64)
+	if err != nil && vb != 0 {
+		writeDevices(w, r, "unable to parse battery voltage: %v", err)
+		return
+	}
+	vnw, err := strconv.ParseFloat(r.FormValue("vnw"), 64)
+	if err != nil && vnw != 0 {
+		writeDevices(w, r, "unable to parse network voltage: %v", err)
+		return
+	}
+	vp1, err := strconv.ParseFloat(r.FormValue("vp1"), 64)
+	if err != nil && vp1 != 0 {
+		writeDevices(w, r, "unable to parse power 1 voltage: %v", err)
+		return
+	}
+	vp2, err := strconv.ParseFloat(r.FormValue("vp2"), 64)
+	if err != nil && vp2 != 0 {
+		writeDevices(w, r, "unable to parse power 2 voltage: %v", err)
+		return
+	}
+	vp3, err := strconv.ParseFloat(r.FormValue("vp3"), 64)
+	if err != nil && vp3 != 0 {
+		writeDevices(w, r, "unable to parse power 3 voltage: %v", err)
+		return
+	}
+	va, err := strconv.ParseFloat(r.FormValue("va"), 64)
+	if err != nil && va != 0 {
+		writeDevices(w, r, "unable to parse alarm voltage: %v", err)
+		return
+	}
+	vr, err := strconv.ParseFloat(r.FormValue("vr"), 64)
+	if err != nil && vr != 0 {
+		writeDevices(w, r, "unable to parse alarm recovery voltage: %v", err)
+		return
+	}
+
+	device, err := model.GetDevice(ctx, settingsStore, model.MacEncode(mac))
+	if err != nil {
+		writeDevices(w, r, "unable to get device to calibrate (%s): %v", mac, err)
+	}
+
+	// Names of the voltage sensors to calibrate.
+	var voltageSensors = []string{
+		model.NameBatterySensor, model.NameNWVoltage, model.NameP1Voltage,
+		model.NameP2Voltage, model.NameP3Voltage,
+	}
+
+	// Load the most recent sensor values.
+	sensors, err := model.GetSensorsV2(ctx, settingsStore, model.MacEncode(mac))
+	if err != nil {
+		writeDevices(w, r, "unable to get sensors (%s): %v", mac, err)
+		return
+	}
+
+	// Calibrate each of the voltage sensors.
+	var msgs []string
+	for _, sensor := range sensors {
+		if sliceutils.ContainsString(voltageSensors, sensor.Name) {
+			scalar, err := model.GetLatestScalar(ctx, mediaStore, model.ToSID(mac, sensor.Pin))
+			if err != nil {
+				msgs = append(msgs, fmt.Sprintf("unable to get latest scalar for %s: %v", sensor.Name, err))
+				continue
+			}
+			reportedTime := time.Unix(scalar.Timestamp, 0)
+
+			// Check if the scalar was recently reported (last 2 monitor periods).
+			if reportedTime.Before(time.Now().Add(-2 * time.Duration(device.MonitorPeriod) * time.Second)) {
+				msgs = append(msgs, fmt.Sprintf("scalar (%s) is out of date (timestamp: %s)(current time: %s)",
+					sensor.Name, reportedTime.Format(time.ANSIC), time.Now().Format(time.ANSIC)))
+				continue
+			}
+
+			var actual float64
+			switch sensor.Name {
+			case model.NameBatterySensor:
+				actual = vb
+			case model.NameNWVoltage:
+				actual = vnw
+			case model.NameP1Voltage:
+				actual = vp1
+			case model.NameP2Voltage:
+				actual = vp2
+			case model.NameP3Voltage:
+				actual = vp3
+			default:
+				// This shouldn't be possible with the ContainsString check.
```

Comment:
If this isn't possible we should probably panic to be loud

Reply:
We should probably have a global panic handler if we don't have one yet, which would allow us to recover from it and notify through email like otv